src/Training.AirBnb.Clone.Backend/Backend Project/Domain/Services/UserService.cs

```diff
@@ -0,0 +1,102 @@
+﻿
+using Backend_Project.Domain.Entities;
+using Backend_Project.Domain.Interfaces;
+using Backend_Project.Persistance.DataContexts;
+using System.Linq.Expressions;
+
+namespace Backend_Project.Domain.Services;
+
+public class UserService : IEntityBaseService<User>
+{
+    private readonly IDataContext _appDataContext;
+    private readonly IValidationService _validationService;
+    public UserService(IDataContext appDataContext, IValidationService validationService)
+    {
+        _appDataContext = appDataContext;
+        _validationService = validationService;
+    }
+
+    public async ValueTask<User> CreateAsync(User entity, bool saveChanges = true)
+    {
+        if (!(await _validationService.IsValidName(entity.FirstName)))
+            throw new FormatException("Invalid first name");
+        if (!(await _validationService.IsValidName(entity.FirstName)))
+            throw new FormatException("Invalid last name");
+        if (!_validationService.IsValidEmailAddress(entity.EmailAddress))
+            throw new FormatException("Invalid email address");
+        if (!(await IsUnique(entity.EmailAddress)))
+            throw new ArgumentException("This email address already exists");
+        await _appDataContext.Users.AddAsync(entity);
+
+        if (saveChanges)
+            await _appDataContext.SaveChangesAsync();
+
+        return entity;
+    }
+
+    public async ValueTask<User> DeleteAsync(Guid id, bool saveChanges = true)
+    {
+        var deletedUser = await GetById(id);
+        if (deletedUser is null)
+            throw new InvalidOperationException("User not found");
+        await _appDataContext.Users.RemoveAsync(deletedUser);
+        deletedUser.DeletedDate = DateTimeOffset.UtcNow;
+        deletedUser.IsDeleted = true;
+        if (saveChanges)
+            await _appDataContext.SaveChangesAsync();
+        return deletedUser;
+
+    }
+
+    public async ValueTask<User> DeleteAsync(User entity, bool saveChanges = true)
+    {
```

Comment:
Bu method ichida ham oldingi hatoliklar takrorlangan shularni to'g'irlab qo'ying.